torch/distributions/kl.py

```diff
@@ -172,14 +172,12 @@ def _kl_beta_beta(p, q):
 @register_kl(Dirichlet, Dirichlet)
 def _kl_dirichlet_dirichlet(p, q):
     # From http://bariskurt.com/kullback-leibler-divergence-between-two-dirichlet-and-beta-distributions/
-    sum_p_alpha = p.alpha.sum(-1)
-    sum_q_alpha = q.alpha.sum(-1)
-    t1 = torch.lgamma(sum_p_alpha)
-    t2 = torch.lgamma(sum_q_alpha)
-    t3 = p.alpha.lgamma().sum(-1)
-    t4 = q.alpha.lgamma().sum(-1)
-    t5 = (p.alpha - q.alpha) * (p.alpha.digamma() - sum_p_alpha.digamma())
-    return t1 - t3 - t2 + t4 + t5.sum(-1)
+    sum_p_alpha = p.alpha.sum(-1).expand_as(p.alpha)
```

Comment:
Don't `.expand_as()`, that's causing a bug. Also you can improve numerically stability by rearranging the first two terms:
```py
p_alpha_sum = p.alpha.sum(-1)
t1 = p_alpha_sum.lgamma() - q.alpha.sum().lgamma()
t2 = (q.alpha.lgamma() - p.alpha.lgamma()).sum(-1)  # sum-after-difference is more numerically stable
t3 = (p.alpha - q.alpha) * (p.alpha.digamma() - p_alpha_sum.digamma().unsqueeze(-1))
return t1 - t2 + t3.sum(-1)
```